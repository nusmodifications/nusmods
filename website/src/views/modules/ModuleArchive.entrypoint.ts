import type { Params } from 'react-router';
import { fetchModuleArchive } from 'actions/moduleBank';
import { captureException } from 'utils/error';
import { Resource, createResource } from 'utils/Resource';
import { JSResource } from 'utils/JSResource';
import type { Module, ModuleCode } from 'types/modules';
import type { EntryPoint } from 'views/routes/types';

type ModuleArchiveResource = Resource<
  { moduleCode: ModuleCode; archiveYear: string },
  string,
  Module
>;

export type PreparedProps = {
  moduleResource: ModuleArchiveResource;
  moduleCode: ModuleCode;
  archiveYear: string;
};

const getPropsFromParams = (params: Params) => {
  const { archiveYear = '', moduleCode = '' } = params;
  return {
    moduleCode: moduleCode.toUpperCase(),
    archiveYear: archiveYear.replace('-', '/'),
  };
};

let moduleResource: ModuleArchiveResource;

/**
 * This is very similar to ModulePage.entrypoint except it is used for the
 * archive page, so it uses different code paths for data fetching.
 */
const entryPoint: EntryPoint<PreparedProps> = {
  component: JSResource(
    'ModuleArchiveContainer',
    () => import(/* webpackChunkName: "ModuleArchiveContainer" */ './ModuleArchiveContainer'),
  ),
  getPreparedProps(params, dispatch) {
    if (!moduleResource) {
      moduleResource = createResource(
        ({ moduleCode, archiveYear }) =>
          dispatch<Module>(fetchModuleArchive(moduleCode, archiveYear)).catch((error) => {
            captureException(error);
            // TODO: If there is an error but module data can still be found, we
            // can assume module has been loaded at some point, so we can just show
            // that instead
            throw error;
          }),
        ({ moduleCode, archiveYear }) =>
          `ModuleArchiveContainer-module-${moduleCode}-${archiveYear}`,
      );
    }

    const { moduleCode, archiveYear } = getPropsFromParams(params);
    moduleResource.preload({ moduleCode, archiveYear });
    return {
      moduleResource,
      moduleCode,
      archiveYear,
    };
  },
  disposePreparedProps(params) {
    const { moduleCode, archiveYear } = getPropsFromParams(params);
    moduleResource.invalidate({ moduleCode, archiveYear });
  },
};

export default entryPoint;
