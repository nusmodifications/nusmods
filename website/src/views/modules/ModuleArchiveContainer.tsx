import { ComponentType, FC, useCallback, useEffect, useRef, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { match as Match, Redirect, useLocation, useRouteMatch } from 'react-router-dom';
import { get } from 'lodash';

import type { AxiosError } from 'axios';
import type { ModuleCode } from 'types/modules';
import type { Dispatch } from 'types/redux';
import type { State } from 'types/state';

import { fetchModuleArchive } from 'actions/moduleBank';
import { captureException } from 'utils/error';
import retryImport from 'utils/retryImport';
import ApiError from 'views/errors/ApiError';
import ModuleNotFoundPage from 'views/errors/ModuleNotFoundPage';
import LoadingSpinner from 'views/components/LoadingSpinner';
import deferComponentRender from 'views/hocs/deferComponentRender';
import { moduleArchive } from 'views/routes/paths';

import type { Props as ModulePageContentProps } from './ModulePageContent';

type Params = {
  year: string;
  moduleCode: string;
};

function getPropsFromMatch(match: Match<Params>) {
  const { year = '', moduleCode = '' } = match.params;
  return {
    moduleCode: moduleCode.toUpperCase(),
    year: year.replace('-', '/'),
  };
}

/**
 * Wrapper component for the archive page that handles data fetching and error handling.
 * This is very similar to ModulePageContainer except it is used for the archive
 * page, so it uses different code paths for canonical URL, data fetching and
 * error handling - the normal page tries to check the archives if this year's
 * API returns 404, while this page doesn't.
 */
export const ModuleArchiveContainerComponent: FC = () => {
  const [ModulePageContent, setModulePageContent] =
    useState<ComponentType<ModulePageContentProps> | null>(null);
  const [error, setError] = useState<Error | undefined>();

  const dispatch = useDispatch<Dispatch>();

  const handleFetchError = useCallback((fetchError: AxiosError) => {
    setError(fetchError);
    captureException(fetchError);
  }, []);

  useEffect(
    () => {
      // Try importing ModulePageContent thrice if we're online and
      // getting the "Loading chunk x failed." error.
      retryImport(() => import(/* webpackChunkName: "module" */ 'views/modules/ModulePageContent'))
        .then((module) => setModulePageContent(() => module.default))
        .catch(handleFetchError);
    },
    // Only load dynamic import on first mount.
    // Assume `handleFetchError` will never change.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [],
  );

  const match = useRouteMatch<Params>();
  const { moduleCode, year } = getPropsFromMatch(match);

  // Fetch module on mount and when params are updated
  const fetchModuleForDisplay = useCallback(() => {
    setError(undefined);
    dispatch(fetchModuleArchive(moduleCode, year)).catch(handleFetchError);
  }, [dispatch, handleFetchError, moduleCode, year]);

  const previousModuleCode = useRef<ModuleCode>();
  const previousArchiveYear = useRef<string>();
  useEffect(() => {
    if (previousModuleCode.current !== moduleCode || previousArchiveYear.current !== year) {
      previousModuleCode.current = moduleCode;
      previousArchiveYear.current = year;
      fetchModuleForDisplay();
    }
  }, [fetchModuleForDisplay, moduleCode, year]);

  const location = useLocation();

  const module = useSelector(({ moduleBank }: State) =>
    get(moduleBank.moduleArchive, [moduleCode, year], null),
  );

  if (get(error, ['response', 'status'], 200) === 404) {
    return <ModuleNotFoundPage moduleCode={moduleCode} tryArchive={false} />;
  }

  // If there is an error but module data can still be found, we assume module has
  // been loaded at some point, so we just show that instead
  if (error && !module) {
    return <ApiError dataName="course information" retry={fetchModuleForDisplay} />;
  }

  // Redirect to canonical URL
  if (module) {
    const canonicalUrl = moduleArchive(moduleCode, year, module.title);
    if (match.url !== canonicalUrl) {
      return <Redirect to={{ ...location, pathname: canonicalUrl }} />;
    }
  }

  if (module && ModulePageContent) {
    // Unique key forces component to remount whenever the user moves to
    // a new module. This allows the internal state (eg. currently selected
    // timetable semester) of <ModulePageContent> to be consistent
    return <ModulePageContent key={`${year}-${moduleCode}`} archiveYear={year} module={module} />;
  }

  return <LoadingSpinner />;
};

export default deferComponentRender(ModuleArchiveContainerComponent);
