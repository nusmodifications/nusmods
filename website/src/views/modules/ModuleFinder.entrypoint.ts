import { JSResource } from 'utils/JSResource';
import { EntryPoint } from 'views/routes/types';

export type PreparedProps = unknown;

const entryPoint: EntryPoint<PreparedProps> = {
  component: JSResource(
    'ModuleFinderContainer',
    () => import(/* webpackChunkName: "ModuleFinderContainer" */ './ModuleFinderContainer'),
  ),
  getPreparedProps() {
    return {};
  },
};

export default entryPoint;
