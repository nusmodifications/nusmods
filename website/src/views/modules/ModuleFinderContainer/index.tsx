import Loadable, { LoadingComponentProps } from 'react-loadable';

import LoadingSpinner from 'views/components/LoadingSpinner';
import ApiError from 'views/errors/ApiError';
import { retryImport } from 'utils/error';

const AsyncModuleFinderContainer = Loadable({
  loader: () =>
    retryImport(() => import(/* webpackChunkName: "module-finder" */ './ModuleFinderContainer')),
  loading: (props: LoadingComponentProps) => {
    if (props.error) {
      return <ApiError dataName="page" retry={props.retry} />;
    }

    if (props.pastDelay) {
      return <LoadingSpinner />;
    }

    return null;
  },
});

export default AsyncModuleFinderContainer;
