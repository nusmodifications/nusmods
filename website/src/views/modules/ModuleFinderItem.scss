@import '~styles/utils/modules-entry';

.modulesItem {
  margin-bottom: 3rem;

  .modulesTitle {
    margin-top: 0.15rem; // align baseline with semester switcher
    font-size: $font-size-lg;
  }

  .notOffered {
    display: flex;
    align-items: center;
    line-height: 1.3;

    .archiveIcon {
      flex: 0 0 1.4rem;
      margin-right: 0.8rem;
    }

    p {
      margin-bottom: 0;
    }
  }
}
