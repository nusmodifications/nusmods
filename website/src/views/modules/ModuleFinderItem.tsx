import * as React from 'react';
import { Link } from 'react-router-dom';
import classnames from 'classnames';

import { ModuleInformation } from 'types/modules';
import { ElasticSearchResult } from 'types/vendor/elastic-search';

import { modulePage } from 'views/routes/paths';
import ModuleSemesterInfo from 'views/components/module-info/ModuleSemesterInfo';
import ModuleWorkload from 'views/components/module-info/ModuleWorkload';
import LinkModuleCodes from 'views/components/LinkModuleCodes';
import { Archive } from 'react-feather';
import { BULLET } from 'utils/react';
import { intersperse } from 'utils/array';
import { mergeModuleHighlight } from 'utils/elasticSearch';
import { isOffered } from 'utils/modules';

import styles from './ModuleFinderItem.scss';

type Props = {
  module: ModuleInformation;
  highlight: ElasticSearchResult<ModuleInformation>['highlight'];
};

// We're setting HTML directly because ElasticSearch returns highlights directly as strings
/* eslint-disable react/no-danger */

const ModuleFinderItem: React.FC<Props> = ({ module, highlight = {} }) => (
  <li className={styles.modulesItem}>
    <div className="row">
      <div className="col-lg-8 col-md-12 col-sm-8">
        <header>
          <h2 className={styles.modulesTitle}>
            <Link to={modulePage(module.moduleCode, module.title)}>
              <span
                dangerouslySetInnerHTML={mergeModuleHighlight(
                  module.moduleCode,
                  highlight.moduleCode,
                )}
              />{' '}
              <span dangerouslySetInnerHTML={mergeModuleHighlight(module.title, highlight.title)} />
            </Link>
          </h2>
          <p>
            {intersperse(
              [
                <span key="department">{module.department}</span>,
                <span key="mc">{module.moduleCredit} MCs</span>,
              ],
              BULLET,
            )}
          </p>
        </header>
        {module.description && (
          <p
            dangerouslySetInnerHTML={mergeModuleHighlight(
              module.description,
              highlight.description,
            )}
          />
        )}
        <dl>
          {module.preclusion && (
            <>
              <dt>Preclusions</dt>
              <dd>
                <LinkModuleCodes>{module.preclusion}</LinkModuleCodes>
              </dd>
            </>
          )}

          {module.prerequisite && (
            <>
              <dt>Prerequisite</dt>
              <dd>
                <LinkModuleCodes>{module.prerequisite}</LinkModuleCodes>
              </dd>
            </>
          )}

          {module.corequisite && (
            <>
              <dt>Corequisite</dt>
              <dd>
                <LinkModuleCodes>{module.corequisite}</LinkModuleCodes>
              </dd>
            </>
          )}
        </dl>
      </div>
      <div className="col-lg-4 col-md-12 col-sm-4">
        {isOffered(module) ? (
          <ModuleSemesterInfo semesters={module.semesterData} moduleCode={module.moduleCode} />
        ) : (
          <div className={classnames(styles.notOffered, 'alert alert-warning')}>
            <Archive className={styles.archiveIcon} />
            <p>This module is not offered this year</p>
          </div>
        )}
        {module.workload && <ModuleWorkload workload={module.workload} />}
      </div>
    </div>
  </li>
);

export default ModuleFinderItem;
