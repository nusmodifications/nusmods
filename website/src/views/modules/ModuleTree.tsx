import * as React from 'react';
import classnames from 'classnames';
import { flatten, values } from 'lodash';
import LinkModuleCodes from 'views/components/LinkModuleCodes';

import { ModuleCode, PrereqTree } from 'types/modules';

import { notNull } from 'types/utils';
import styles from './ModuleTree.scss';

type Props = {
  moduleCode: ModuleCode;
  fulfillRequirements?: readonly ModuleCode[];
  prereqTree?: PrereqTree;
};

interface TreeDisplay {
  layer: number;
  node: PrereqTree;
  isPrereq?: boolean;
}

const GRADE_REQUIREMENT_SEPARATOR = ':';
const MODULE_NAME_WILDCARD = '%';
const PASSING_GRADE = 'D';

const formatConditional = (node: PrereqTree) => {
  if (typeof node === 'string') return node;
  if ('nOf' in node) {
    const requiredNum = node.nOf[0];
    return `at least ${requiredNum} of`;
  }
  if ('or' in node) {
    return 'one of';
  }
  return 'all of';
};

const nodeName = (node: PrereqTree) => {
  if (typeof node !== 'string') {
    return Object.keys(node)[0];
  }
  let name = node;
  if (name.includes(GRADE_REQUIREMENT_SEPARATOR)) {
    const [moduleName, requiredGrade] = name.split(GRADE_REQUIREMENT_SEPARATOR);
    if (requiredGrade !== PASSING_GRADE) {
      name = `${moduleName} (minimally ${requiredGrade})`;
    } else {
      name = moduleName;
    }
  }
  if (name.includes(MODULE_NAME_WILDCARD)) {
    const [beforeWildcard, afterWildcard] = name.split(MODULE_NAME_WILDCARD);
    name = `course starting with '${beforeWildcard}' ${afterWildcard}`;
  }
  return name.trim();
};

const unwrapLayer = (node: PrereqTree) => {
  if (typeof node === 'string') {
    return [node];
  }
  if ('nOf' in node) {
    return node.nOf[1];
  }
  return flatten(values(node).filter(notNull));
};

const Branch: React.FC<{ nodes: PrereqTree[]; layer: number }> = (props) => (
  <ul className={styles.tree}>
    {props.nodes.map((child, idx) => (
      <li className={styles.branch} key={typeof child === 'string' ? nodeName(child) : idx}>
        <Tree node={child} layer={props.layer} />
      </li>
    ))}
  </ul>
);

const Tree: React.FC<TreeDisplay> = (props) => {
  const { layer, node, isPrereq } = props;

  const isConditional = typeof node !== 'string';
  const name = nodeName(node);

  return (
    <>
      <div
        className={classnames(styles.node, {
          [`hoverable color-${layer}`]: !isConditional,
          [styles.conditional]: isConditional,
          [styles.prereqNode]: isPrereq,
        })}
      >
        {isConditional ? (
          formatConditional(node)
        ) : (
          <LinkModuleCodes className={styles.link}>{name}</LinkModuleCodes>
        )}
      </div>

      {isConditional && <Branch nodes={unwrapLayer(node)} layer={layer + 1} />}
    </>
  );
};

const ModuleTree: React.FC<Props> = (props) => {
  const { fulfillRequirements, prereqTree, moduleCode } = props;

  return (
    <>
      <div className={styles.container}>
        {fulfillRequirements && fulfillRequirements.length > 0 && (
          <>
            <ul className={styles.prereqTree}>
              {fulfillRequirements.map((fulfilledModule) => (
                <li
                  key={fulfilledModule}
                  className={classnames(styles.branch, styles.prereqBranch)}
                >
                  <Tree layer={0} node={fulfilledModule} isPrereq />
                </li>
              ))}
            </ul>

            <div className={classnames(styles.node, styles.conditional)}>needs</div>
          </>
        )}

        <ul className={classnames(styles.tree, styles.root)}>
          <li className={classnames(styles.branch)}>
            <Tree layer={1} node={moduleCode} />

            {prereqTree && <Branch nodes={[prereqTree]} layer={2} />}
          </li>
        </ul>
      </div>

      {/* <p className="alert alert-warning">
        The prerequisite tree is displayed for visualization purposes and may not be accurate.
        Viewers are encouraged to double check details.
      </p> */}

      <p className="alert alert-warning">
        This new version of the prerequisite tree is being tested and may not be accurate. Viewers
        are encouraged to double check details with the prerequisite text above. To report bugs with
        the new tree, please post a bug report on GitHub (preferred) at{' '}
        <a
          href="https://github.com/nusmodifications/nusmods/issues/new/choose"
          target="_blank"
          rel="noopener noreferrer nofollow"
        >
          our repository
        </a>{' '}
        or send an email to{' '}
        <a href="mailto:bugs@nusmods.com" target="_blank" rel="noopener noreferrer nofollow">
          bugs@nusmods.com
        </a>
        .
      </p>
    </>
  );
};

export default ModuleTree;
