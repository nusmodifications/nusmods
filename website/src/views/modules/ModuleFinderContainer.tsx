import React from 'react';
import {
  Hits,
  HitsStats,
  HitsStatsDisplayProps,
  InitialLoader,
  LoadingComponent,
  HitsListProps,
  NoHits,
  Pagination,
  SearchkitManager,
  SearchkitProvider,
} from 'searchkit';
import classnames from 'classnames';

import { ElasticSearchResult } from 'types/vendor/elastic-search';
import { ModuleInformation } from 'types/modules';

import ModuleFinderSidebar from 'views/modules/ModuleFinderSidebar';
import ModuleSearchBox from 'views/modules/ModuleSearchBox';
import ModuleFinderNoHits from 'views/errors/ModuleFinderNoHits';
import ModuleFinderApiError from 'views/errors/ModuleFinderApiError';
import ModuleFinderPager from 'views/components/ModuleFinderPager';
import ModuleFinderItem from 'views/components/ModuleFinderItem';
import LoadingSpinner from 'views/components/LoadingSpinner';
import Title from 'views/components/Title';

import { forceElasticsearchHost } from 'utils/debug';
import { HIGHLIGHT_OPTIONS } from 'utils/elasticSearch';
import config from 'config';
import styles from './ModuleFinderContainer.scss';

const esHostUrl = `${forceElasticsearchHost() || config.elasticsearchBaseUrl}/modules`;
const searchkit = new SearchkitManager(esHostUrl);

const pageHead = <Title>Modules</Title>;

/* eslint-disable no-underscore-dangle */

const ModuleInformationListComponent: React.FC<HitsListProps> = ({ hits }) => (
  <ul className={styles.modulesList}>
    {hits.map((hit) => {
      const result = hit as ElasticSearchResult<ModuleInformation>;
      return (
        <ModuleFinderItem
          key={result._source.moduleCode}
          module={result._source}
          highlight={result.highlight}
        />
      );
    })}
  </ul>
);

const ModuleFinderContainer: React.FC = () => {
  return (
    <div className={classnames(styles.modulesPageContainer, 'page-container')}>
      {pageHead}
      <SearchkitProvider searchkit={searchkit}>
        <div className="row">
          <div className="col">
            <h1 className="sr-only">Module Finder</h1>

            <ModuleSearchBox id="q" />

            <div>
              <HitsStats
                component={({ hitsCount }: HitsStatsDisplayProps) =>
                  hitsCount > 0 ? (
                    <div className={styles.modulePageDivider}>
                      {hitsCount} modules found{' '}
                      <LoadingComponent>
                        <LoadingSpinner small />
                      </LoadingComponent>
                    </div>
                  ) : (
                    <LoadingComponent>
                      <LoadingSpinner />
                    </LoadingComponent>
                  )
                }
              />

              <InitialLoader component={LoadingComponent} />

              <Hits
                hitsPerPage={10}
                listComponent={ModuleInformationListComponent}
                customHighlight={HIGHLIGHT_OPTIONS}
                // SearchKit default incorrectly tries to set document.body.scrollTop. <html> is the correct
                // scrolling context for the viewport
                scrollTo="html"
              />

              <NoHits
                suggestionsField="title"
                component={ModuleFinderNoHits}
                errorComponent={ModuleFinderApiError}
              />
            </div>

            <Pagination showNumbers listComponent={ModuleFinderPager} />
          </div>

          <div className="col-md-4 col-lg-3">
            <ModuleFinderSidebar />
          </div>
        </div>
      </SearchkitProvider>
    </div>
  );
};

export default ModuleFinderContainer;
