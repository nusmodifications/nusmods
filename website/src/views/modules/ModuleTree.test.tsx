import { render } from 'enzyme';

import { getModuleCondensed } from 'selectors/moduleBank';
import { ModuleCondensed } from 'types/modules';
import { ModuleTreeComponent } from './ModuleTree';

jest.mock('views/components/LinkModuleCodes', () => 'mockedlink');

describe(ModuleTreeComponent, () => {
  test('should render requirements fulfilled tree of module', () => {
    const component = render(
      <ModuleTreeComponent
        moduleCode="ACC1002"
        getModuleCondensed={getModuleCondensed({ moduleBank: { moduleCodes: {} } } as any)}
        fulfillRequirements={[
          'ACC1006',
          'ACC2002',
          'ACC3601',
          'ACC3603',
          'ACC3605',
          'ACC3616',
          'FIN2004',
          'FIN2004X',
          'FIN3113',
          'FIN3130',
          'IS5116',
          'ACC3611',
          'FIN3132',
          'FIN4115',
        ]}
      />,
    );

    expect(component).toMatchSnapshot('ACC1002');
  });

  test('should render prereq tree of module', () => {
    const component = render(
      <ModuleTreeComponent
        moduleCode="CS3244"
        getModuleCondensed={getModuleCondensed({ moduleBank: { moduleCodes: {} } } as any)}
        fulfillRequirements={['CS5242', 'CS5339', 'CS6281']}
        prereqTree={{
          and: [
            {
              or: ['CS2010', 'CS2020', 'CS2040', 'CS2040C'],
            },
            {
              or: ['ESP1107', 'ESP2107', 'ST1232', 'ST2131', 'ST2132', 'ST2334'],
            },
            {
              or: ['MA1101R', 'MA1311', 'MA1506'],
            },
            {
              or: ['MA1102R', 'MA1505', 'MA1521'],
            },
          ],
        }}
      />,
    );

    expect(component).toMatchSnapshot('CS3244');
  });

  // Test that modules which are in moduleBank have appropriate colours,
  // and modules that aren't are greyed out

  const testModules: { [moduleCode: string]: ModuleCondensed } = {
    CS2040: {
      moduleCode: 'CS2040',
      title: 'Data Structures and Algorithms',
      semesters: [1, 2],
    },
    CS2030: {
      moduleCode: 'CS2030',
      title: 'Programming Methodology II',
      semesters: [1, 2],
    },
    CS2113T: {
      moduleCode: 'CS2113T',
      title: 'Software Engineering & Object-Oriented Programming',
      semesters: [1, 2],
    },
    CS1020E: {
      moduleCode: 'CS1020E',
      title: 'Data Structures and Algorithms',
      semesters: [1, 2],
    },
    CS6240: {
      moduleCode: 'CS6240',
      title: 'Multimedia Analysis',
      semesters: [2],
    },
  };

  test('should grey out modules that are not in module bank', () => {
    const component = render(
      <ModuleTreeComponent
        moduleCode="CS4243"
        getModuleCondensed={getModuleCondensed({ moduleBank: { moduleCodes: testModules } } as any)}
        fulfillRequirements={['CS6240', 'CS3281', 'CS4243R']}
        prereqTree={{
          and: [
            {
              or: [
                'CS1020',
                'CS1020E',
                'CS2020',
                {
                  and: [
                    {
                      or: ['CS2030', 'CS2113', 'CS2113T'],
                    },
                    {
                      or: ['CS2040', 'CS2040C'],
                    },
                  ],
                },
              ],
            },
          ],
        }}
      />,
    );

    expect(component).toMatchSnapshot('CS4243');
  });
});
