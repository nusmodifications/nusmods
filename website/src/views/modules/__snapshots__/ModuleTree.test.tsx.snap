// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleTree should render prereq tree of module: CS3244 1`] = `
[
  <div
    class="container"
  >
    <ul
      class="prereqTree"
    >
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            CS5242
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            CS5339
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            CS6281
          </mockedlink>
        </div>
      </li>
    </ul>
    <div
      class="node conditional"
    >
      needs
    </div>
    <ul
      class="tree root"
    >
      <li
        class="branch"
      >
        <div
          class="node moduleNode hoverable color-1"
        >
          <mockedlink
            class="link"
          >
            CS3244
          </mockedlink>
        </div>
        <ul
          class="tree"
        >
          <li
            class="branch"
          >
            <div
              class="node conditional"
            >
              all of
            </div>
            <ul
              class="tree"
            >
              <li
                class="branch"
              >
                <div
                  class="node conditional"
                >
                  one of
                </div>
                <ul
                  class="tree"
                >
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        CS2010
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        CS2020
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        CS2040
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        CS2040C
                      </mockedlink>
                    </div>
                  </li>
                </ul>
              </li>
              <li
                class="branch"
              >
                <div
                  class="node conditional"
                >
                  one of
                </div>
                <ul
                  class="tree"
                >
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        ESP1107
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        ESP2107
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        ST1232
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        ST2131
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        ST2132
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        ST2334
                      </mockedlink>
                    </div>
                  </li>
                </ul>
              </li>
              <li
                class="branch"
              >
                <div
                  class="node conditional"
                >
                  one of
                </div>
                <ul
                  class="tree"
                >
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        MA1101R
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        MA1311
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        MA1506
                      </mockedlink>
                    </div>
                  </li>
                </ul>
              </li>
              <li
                class="branch"
              >
                <div
                  class="node conditional"
                >
                  one of
                </div>
                <ul
                  class="tree"
                >
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        MA1102R
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        MA1505
                      </mockedlink>
                    </div>
                  </li>
                  <li
                    class="branch"
                  >
                    <div
                      class="node moduleNode hoverable color-4"
                    >
                      <mockedlink
                        class="link"
                      >
                        MA1521
                      </mockedlink>
                    </div>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </div>,
  <p
    class="alert alert-warning"
  >
    This new version of the prerequisite tree is being tested and may not be accurate. Viewers are encouraged to double check details with the prerequisite text above. To report bugs with the new tree, please post a bug report on GitHub (preferred) at 
    <a
      href="https://github.com/nusmodifications/nusmods/issues/new/choose"
      rel="noopener noreferrer nofollow"
      target="_blank"
    >
      our repository
    </a>
     or send an email to 
    <a
      href="mailto:bugs@nusmods.com"
      rel="noopener noreferrer nofollow"
      target="_blank"
    >
      bugs@nusmods.com
    </a>
    .
  </p>,
]
`;

exports[`ModuleTree should render requirements fulfilled tree of module: ACC1002 1`] = `
[
  <div
    class="container"
  >
    <ul
      class="prereqTree"
    >
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            ACC1006
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            ACC2002
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            ACC3601
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            ACC3603
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            ACC3605
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            ACC3616
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            FIN2004
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            FIN2004X
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            FIN3113
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            FIN3130
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            IS5116
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            ACC3611
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            FIN3132
          </mockedlink>
        </div>
      </li>
      <li
        class="branch prereqBranch"
      >
        <div
          class="node moduleNode hoverable color-0 prereqNode"
        >
          <mockedlink
            class="link"
          >
            FIN4115
          </mockedlink>
        </div>
      </li>
    </ul>
    <div
      class="node conditional"
    >
      needs
    </div>
    <ul
      class="tree root"
    >
      <li
        class="branch"
      >
        <div
          class="node moduleNode hoverable color-1"
        >
          <mockedlink
            class="link"
          >
            ACC1002
          </mockedlink>
        </div>
      </li>
    </ul>
  </div>,
  <p
    class="alert alert-warning"
  >
    This new version of the prerequisite tree is being tested and may not be accurate. Viewers are encouraged to double check details with the prerequisite text above. To report bugs with the new tree, please post a bug report on GitHub (preferred) at 
    <a
      href="https://github.com/nusmodifications/nusmods/issues/new/choose"
      rel="noopener noreferrer nofollow"
      target="_blank"
    >
      our repository
    </a>
     or send an email to 
    <a
      href="mailto:bugs@nusmods.com"
      rel="noopener noreferrer nofollow"
      target="_blank"
    >
      bugs@nusmods.com
    </a>
    .
  </p>,
]
`;
